polynote-runtime/src/main/scala/polynote/runtime/ValueRepr.scala

```diff
@@ -0,0 +1,249 @@
+package polynote.runtime
+
+import java.nio.ByteBuffer
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+import sun.misc.Cleaner
+
+/**
+  * A plain text representation of a value
+  */
+final case class StringRepr(string: String) extends ValueRepr
+
+/**
+  * A MIME representation of a value. Binary values must be base64-encoded.
+  */
+final case class MIMERepr(mimeType: String, content: String) extends ValueRepr
+
+/**
+  * A binary representation of a value, encoded in Polynote's format (TODO)
+  */
+final case class DataRepr(dataType: DataType, data: ByteBuffer) extends ValueRepr
+
+/**
+  * An identifier for a "lazy" value. It won't be evaluated until the client asks for it by its handle.
+  * We use Ints for the handle, because JS needs to store the handle, and it can't store Long. We just can't have
+  * more than 2^32^ different handles... I think there would be bigger issues at that point.
+  */
+final case class LazyDataRepr private[polynote](handle: Int, dataType: DataType) extends ValueRepr {
+  def isEvaluated: Option[Boolean] = LazyDataRepr.getHandle(handle).map(_.isEvaluated)
+
+  def release(): Unit = {
+    LazyDataRepr.handles.remove(handle)
+  }
+}
+
+object LazyDataRepr {
+
+  private[polynote] class Handle(val handle: Int, val dataType: DataType, lazyData: => ByteBuffer) {
+    @volatile private var computedFlag: Boolean = false
+
+    lazy val data: ByteBuffer = {
+      computedFlag = true
+      lazyData
+    }
+
+    def isEvaluated: Boolean = computedFlag
+  }
+
+
+  private val handles: ConcurrentHashMap[Int, Handle] = new ConcurrentHashMap()
+  private val nextHandle: AtomicInteger = new AtomicInteger(0)
+  private[polynote] def getHandle(handleId: Int): Option[Handle] = Option(handles.get(handleId))
+
+  def apply(dataType: DataType, value: => ByteBuffer): LazyDataRepr = {
+    val handleId = nextHandle.getAndIncrement()
+    val handle = new Handle(handleId, dataType, value)
+    handles.put(handleId, handle)
+    val repr = LazyDataRepr(handleId, dataType)
+
+    // if the repr object gets GC'ed, we can let the handle go as well
```

Comment:
Is it possible that the UI might request data for a handle that has been GCed? I guess that since Repr values eventually get held by the NotebookCell, they would only get GCed if we unloaded the notebook for some reason (which I don't think we do now?). At which point the UI wouldn't be asking for that repr so we should be ok.

Reply:
Yeah, I think you're right. If they did, it wouldn't be catastrophic though (just an error that hopefully doesn't terminate stuff!)